AppMakefile.mk

```diff
@@ -45,6 +45,15 @@ include $(TOCK_USERLAND_BASE_DIR)/Program.mk
 
 # Rules to incorporate external libraries
 define EXTERN_LIB_RULES
+
+# If there is a Makefile.setup, then we try to automatically build the library.
+ifneq "$$(wildcard $(1)/Makefile.setup)" ""
+  # Do any setup steps needed to fetch the library.
+  DEVNULL := $$(shell make -C $(1) -f Makefile.setup all)
+  # Build the library if needed.
+  DEVNULL := $$(shell make -C $(1) all)
+endif
```

Comment:
This is a red flag. Generally, our build system tries to follow the ethos of [recursive make considered harmful](https://accu.org/journals/overload/14/71/miller_2004/). Everything else carefully stitches together rules so that there is only ever one invocation of the `make` program which knows the full build.

I don't see anything in the `u8g2` example with this PR that couldn't be done as rules. Probably something like have included makefiles add their final targets to a variable like `PREBUILD_DEPENDENCIES`, and then add `PREBUILD_DEPENDENCIES` as another dependency roughly everywhere the `BUILD_DIR` dependency already is.

Reply:
One issue:
1. We need to download a .zip file.
2. That .zip file contains the c files that are picked up by the wildcard.

Is there a way to populate the SRCs variable after a recipe has run?

Reply:
Second issue: we need to build N libraries. We only have one TockLibrary makefile. That makefile uses `$(LIBNAME)`, which can only be set once.

Reply:
However, `$(LIBNAME)` is used in almost every line in that makefile so it's probably gonna get gnarly.

Reply:
While I haven't tested it in a while, the multiple library name issue should already be addressed, that's what https://github.com/tock/libtock-c/blob/master/TockLibrary.mk#L134-L148 is about.

`$(LIBNAME)` is only used while `TockLibrary.mk` is generating rules, so if you redefine `LIBNAME` (must use `:=`) and include `TockLibrary.mk` again, things should work.

Reply:
_Should_, but only if the second library contains no files with the same name as the first library.

Reply:
> Is there a way to populate the SRCs variable after a recipe has run?

This one is a bigger and more fundamental issue. Make expects to first fully define the dependency graph and then start resolving it. It does not, by design, allow for the dependency graph to be modified once it starts running rules.

---

A _hack_ of an answer would be something like

```
LIBNAME_SRCS := $(shell test -d unzipped_dir || (wget ZIPFILE && unzip ZIPFILE) && echo unzipped_dir/*.c)
```

Reply:
This is why all the other libraries that we're interested in building at runtime are submodules; lets version control solve the "get the source files there before running make" problem.

Probably best to do the same for `u8g2`, _or_ to simply commit pre-compiled objects like we do for newlib/etc.